core/code/chat.js

```diff
@@ -501,48 +508,75 @@ window.chat.renderMarkupEntity = function (ent) {
 
 window.chat.renderMarkup = function (markup) {
   var msg = '';
-  markup.forEach(function(ent, ind) {
+
+  markup.forEach(function (ent, ind) {
     switch (ent[0]) {
-    case 'SENDER':
-    case 'SECURE':
-      // skip as already handled
-      break;
+      case 'SENDER':
+      case 'SECURE':
+        // skip as already handled
+        break;
 
-    case 'PLAYER': // automatically generated messages
-      if (ind > 0) msg += chat.renderMarkupEntity(ent); // don’t repeat nick directly
-      break;
+      case 'PLAYER': // automatically generated messages
+        if (ind > 0) msg += chat.renderMarkupEntity(ent); // don’t repeat nick directly
+        break;
 
-    default:
-      // add other enitities whatever the type
-      msg += chat.renderMarkupEntity(ent);
-      break;
+      default:
+        // add other enitities whatever the type
+        msg += chat.renderMarkupEntity(ent);
+        break;
     }
   });
   return msg;
 };
 
-window.chat.renderTimeCell = function(time, classNames) {
-  var ta = unixTimeToHHmm(time);
-  var tb = unixTimeToDateTimeString(time, true);
+function transformMessage(markup) {
+  // Make a copy of the markup array to avoid modifying the original input
+  let newMarkup = [...markup];
+
+  // Collapse <faction> + "Link"/"Field". Example: "Agent <player> destroyed the <faction> Link ..."
+  if (newMarkup.length > 4) {
+    if (newMarkup[3][0] === 'FACTION' && newMarkup[4][0] === 'TEXT' && (newMarkup[4][1].plain === ' Link ' || newMarkup[4][1].plain === ' Control Field @')) {
+      newMarkup[4][1].team = newMarkup[3][1].team;
```

Comment:
this line modifies the original input, this would need a deep copy (basically `JSON.parse(JSON.stringify(...))`)